lib/rust/ensogl/core/src/debug/stats.rs

```diff
@@ -15,48 +15,137 @@
+// ====================
+
+/// An entity that can retrieve current time.
+pub trait TimeProvider {
+    /// Returns current time, measured in milliseconds.
+    fn now(&self) -> f64;
+}
+
+impl TimeProvider for Performance {
+    fn now(&self) -> f64 {
+        self.now()
+    }
+}
+
+
+
 // =============
 // === Stats ===
 // =============
 
-/// Structure containing all the gathered stats.
-#[derive(Debug, Clone, CloneRef)]
-pub struct Stats {
-    rc: Rc<RefCell<StatsData>>,
+/// Tracks stats related to the current rendering frame, providing methods for modifying and
+/// retrieving their values.
+pub type Stats = StatsOverTime<Performance>;
+
+
+// === StatsOverTime ===
+
+/// Tracks stats related to the current rendering frame, providing methods for modifying and
+/// retrieving their values.
+#[derive(Debug, CloneRef)]
+pub struct StatsOverTime<T: TimeProvider> {
+    rc: Rc<RefCell<StatsCollector<T>>>,
 }
 
-impl Default for Stats {
-    fn default() -> Self {
-        let rc = Rc::new(RefCell::new(default()));
-        Self { rc }
+impl<T: TimeProvider> Clone for StatsOverTime<T> {
+    fn clone(&self) -> Self {
+        Self { rc: self.rc.clone() }
     }
 }
 
-impl Stats {
+impl<T: TimeProvider> StatsOverTime<T> {
+    /// Constructor.
+    pub fn new(time_provider: T) -> Self {
+        let stats_collector = StatsCollector::new(time_provider);
+        let rc = Rc::new(RefCell::new(stats_collector));
+        Self { rc }
+    }
+
     /// Starts tracking data for a new animation frame.
-    /// Also, calculates the `fps` stat and updates `frame_counter`.
+    /// Also, calculates the [`fps`] stat.
     /// Returns a snapshot of statistics data for the previous frame.
-    /// Note: on first ever frame, there was no "previous frame", so all returned stats are zero
-    /// (this special case can be recognized by checking `frame_counter == 0`).
-    pub fn begin_frame(&self, time: f64) -> StatsData {
-        self.rc.borrow_mut().begin_frame(time)
+    ///
+    /// Note: on first ever frame, there was no "previous frame", so no data will be returned.
+    ///
+    /// Note: the code works under an assumption that [`begin_frame()`] and [`end_frame()`] are
+    /// called properly on every frame (behavior in case of missed frames or missed calls is not
+    /// specified).
+    pub fn begin_frame(&self) -> Option<StatsData> {
+        self.rc.borrow_mut().begin_frame()
     }
 
     /// Ends tracking data for the current animation frame.
     /// Also, calculates the `frame_time` and `wasm_memory_usage` stats.
-    pub fn end_frame(&self, time: f64) {
-        self.rc.borrow_mut().end_frame(time);
+    pub fn end_frame(&self) {
+        self.rc.borrow_mut().end_frame();
     }
+}
+
+
+
+// ======================
+// === StatsCollector ===
+// ======================
+
+#[derive(Debug)]
+struct StatsCollector<T: TimeProvider> {
+    time_provider:    T,
+    stats_data:       StatsData,
+    frame_begin_time: Option<f64>,
+}
```

Comment:
I'm not sure whether this struct is collecting stats. It has a reference to StatsData (which contains stats) and then it cleans the stats on every frame, and adds info about memory usage and frame time after the frame. I would not name it a "stats collector" in such a case.

Reply:
Alternatives `StatsHandler`.